src/util.cc

```diff
@@ -2428,38 +2462,56 @@ string json_escape_utf8(const string& val)
 		if ( ( ch >= 32 && ch <= 127 ) ||
 		       ch == '\b' || ch == '\f' || ch == '\n' || ch == '\r' || ch == '\t' )
 			{
-			result.push_back(ch);
+			if ( ! found_bad )
+				utf_result.push_back(ch);
+
+			escaped_result.push_back(ch);
 			++idx;
 			continue;
 			}
-		else if ( ch >= 0 && ch < 32 )
+		else if ( found_bad )
 			{
-			result.append(json_escape_byte(ch));
+			// If we already found a bad UTF8 character (see check_ok_utf8) just insert the bytes
+			// as escaped characters into the escaped result and move on.
+			escaped_result.append(json_escape_byte(ch));
 			++idx;
 			continue;
 			}
 
-		// Find out how long the next character should be.
-		unsigned int char_size = getNumBytesForUTF8(ch);
-
-		// If it says that it's a single character or it's not an valid string UTF8 sequence, insert
-		// the one escaped byte into the string, step forward one, and go to the next character.
-		if ( char_size == 0 || idx+char_size > val_size || isLegalUTF8Sequence(val_data+idx, val_data+idx+char_size) == 0 )
+		// If we haven't found a bad UTF-8 character yet, check to see if the next one starts a
+		// UTF-8 character. If not, we'll mark that we're on a bad result. Otherwise we'll go
+		// ahead and insert this character and continue.
+		if ( ! found_bad )
 			{
-			result.append(json_escape_byte(ch));
-			++idx;
-			continue;
-			}
+			// Find out how long the next character should be.
+			unsigned int char_size = getNumBytesForUTF8(ch);
 
-		result.append(val, idx, char_size);
-		idx += char_size;
+			// If it says that it's a single character or it's not an valid string UTF8 sequence, insert
+			// the one escaped byte into the string, step forward one, and go to the next character.
+			if ( char_size == 0 || idx+char_size > val_size ||
+			     ! check_ok_utf8(val_data+idx, val_data+idx+char_size) )
+				{
+				found_bad = true;
+				escaped_result.append(json_escape_byte(ch));
+				++idx;
+				continue;
+				}
+			else
+				{
+				utf_result.append(val, idx, char_size);
+				idx += char_size;
+				}
+			}
 		}
 
 	// Insert any of the remaining bytes into the string as escaped bytes
 	for ( ; idx < val_size; ++idx )
-		result.append(json_escape_byte(val[idx]));
+		escaped_result.append(json_escape_byte(val[idx]));
```

Comment:
Can this loop still happen? The condition is the same as with the previous loop, and that one has no `break`.

Reply:
Oh weird. There wasn't a break there before either. That must have been a leftover artifact from a previous change.